Implement type-checkable handles

## Problem
Handle types are currently implemented as `typedef`s to `void *`, which means that a C compiler cannot do type checking on handles.

For example, here is a C program that uses the C bindings for [Quilc](https://github.com/quil-lang/quilc), violates the function type signatures specified by [`libquilc`](https://github.com/quil-lang/quilc/blob/master/lib/src/libquilc.lisp), but compiles without any issues:

```c
#include <stdio.h>
#include <stdlib.h>

#include "libquilc.h"

void die(char *msg) {
  printf("%s\n", msg);
  exit(1);
}

int main(int argc, char **argv) {
  init("libquilc.core");

  char source[256];
  chip_specification chip_spec;

  fgets(source, 256, stdin);

  /*
   * The function call below should not compile, since the function `quilc_parse_quil` has the signature
   * 
   *     error_t (*quilc_parse_quil)(char* source, quil_program *result)
   * 
   * but we are calling it with a `chip_specification *` as the second argument.
   */
  if (quilc_parse_quil(source, &chip_spec) != ERROR_SUCCESS)
    die("unable to parse quil");

  lisp_release_handle(chip_spec);

  return 0;
}
```


## Solution
Implement handles as `typedef`s to dummy struct pointers. For example, we can redefine the `quil_program` and `chip_specification` handle types as such:

```c
struct quil_program__ { int unused; }; typedef struct quil_program__ *quil_program;
struct chip_specification__ { int unused; }; typedef struct chip_specification__ *chip_specification;
```

Now the example program given above causes the compiler to emit a warning:

```c
test.c:19:32: warning: incompatible pointer types passing 'chip_specification *' (aka 'struct chip_specification__ **') to
      parameter of type 'quil_program *' (aka 'struct quil_program__ **') [-Wincompatible-pointer-types]
  if (quilc_parse_quil(source, &chip_spec) != ERROR_SUCCESS)
                               ^~~~~~~~~~
1 warning generated.
```

Notably, this technique of implementing opaque handles is standard in practice in Win32 [[1](https://stackoverflow.com/questions/376452/enforce-strong-type-checking-in-c-type-strictness-for-typedefs/376478), [2](https://github.com/tpn/winsdk-7/blob/master/v7.1A/Include/WinNT.h#L519)].

--- a/types.lisp
+++ b/types.lisp
@@ -110,7 +110,7 @@ Handles serve roughly as opaque pointers to Lisp objects."
   `(define-type ,name
        :c-type ,c-type
        :alien-type (sb-alien:* t)
-       :definition ,(format nil "typedef void* ~a;" c-type)
+       :definition ,(format nil "struct ~a__ { int unused; }; typedef struct ~:*~a__ *~:*~a;" c-type)
        :python-type ,c-type
        :python-type-definition ,(format nil "class ~a(c_void_p):~%    pass~%~%" c-type)
        :lisp-to-alien (lambda (c) `(make-handle ,c))

